hail/src/main/scala/is/hail/io/fs/GoogleStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+        pos += toTransfer
+        toTransfer
+      }
+
+      def getPosition: Long = pos
+
+      def seek(newPos: Long): Unit = {
+        bb.clear()
+        reader.seek(newPos)
+        pos = newPos
+      }
+    }
+
+    new WrappedSeekableDataInputStream(is)
+  }
+
+  def createNoCompression(filename: String): PositionedDataOutputStream = {
+    val (bucket, path) = getBucketPath(filename)
+
+    val blobId = BlobId.of(bucket, path)
+    val blobInfo = BlobInfo.newBuilder(blobId)
+      .build()
+
+    val os: PositionedOutputStream = new OutputStream with Positioned {
+      val bb: ByteBuffer = ByteBuffer.allocate(64 * 1024)
+      var pos: Long = 0
+      val write: WriteChannel = storage.writer(blobInfo)
+
+      override def flush(): Unit = {
+        bb.flip()
+
+        while (bb.remaining() > 0)
+          write.write(bb)
+
+        bb.clear()
+      }
+
+      override def write(i: Int): Unit = {
+        if (bb.remaining() == 0)
+          flush()
+        bb.put(i.toByte)
+        pos += 1
+      }
+
+      override def write(bytes: Array[Byte], off: Int, len: Int): Unit = {
+        var i = off
+        var remaining = len
+        while (remaining > 0) {
+          if (bb.remaining() == 0)
+            flush()
+          val toTransfer = math.min(bb.remaining(), remaining)
+          bb.put(bytes, i, toTransfer)
+          i += toTransfer
+          remaining -= toTransfer
+          pos += toTransfer
+        }
+      }
+
+      override def close(): Unit = {
+        flush()
+        write.close()
+      }
+
+      def getPosition: Long = pos
+    }
+
+    new WrappedPositionedDataOutputStream(os)
+  }
+
+  def mkDir(dirname: String): Unit = ()
+
+  def delete(filename: String, recursive: Boolean): Unit = {
+    val (bucket, path) = getBucketPath(filename)
+    if (recursive) {
+      val it = storage.list(bucket, BlobListOption.prefix(path))
+        .getValues.iterator.asScala
+      while (it.hasNext) {
+        storage.delete(it.next().getBlobId)
+      }
+    } else
+      storage.delete(bucket, path)
+  }
+
+  def glob(filename: String): Array[FileStatus] = {
+    var (bucket, path) = getBucketPath(filename)
+    path = dropTrailingSlash(path)
+
+    val components =
+      if (path == "")
+        Array.empty[String]
+      else
+        path.split("/")
+
+    val javaFS = FileSystems.getDefault
+
+    val ab = new mutable.ArrayBuffer[FileStatus]()
+    def f(prefix: String, fs: FileStatus, i: Int): Unit = {
+      assert(!prefix.endsWith("/"), prefix)
+
+      if (i == components.length) {
+        var t = fs
+        if (t == null) {
+          try {
+            t = fileStatus(prefix)
+          } catch {
+            case _: FileNotFoundException =>
+          }
+        }
+        if (t != null)
+          ab += t
+      }
+
+      if (i < components.length) {
+        val c = components(i)
+        if (containsWildcard(c)) {
+          val m = javaFS.getPathMatcher(s"glob:$c")
```

Comment:
I would make this a lazy static (object) attribute, drop the containsWildcard(c), and perform `if (m.matches(c)` directly, unless trailing slash rules break the glob matcher.

Reply:
I can't do that because it means I do a directory listing (which can be very expensive) if the component is not a pattern.  See my comments above.